lib/ublksrv_cmd.c

```diff
@@ -160,6 +174,33 @@ int ublksrv_ctrl_get_affinity(struct ublksrv_ctrl_dev *ctrl_dev)
 	return 0;
 }
 
+int ublksrv_ctrl_report_zones_fetch(const struct ublksrv_ctrl_dev *ctrl_dev,
+		struct io_uring *ring,
+		uint64_t *cqe_data) {
+
+	struct ublksrv_ctrl_cmd_data data = {
+		.cmd_op = UBLK_CMD_REPORT_ZONES_FETCH,
+	};
+
+	return __ublksrv_ctrl_cmd_with_ring(ctrl_dev, ring, &data, cqe_data);
+}
+
+int ublksrv_ctrl_report_zones_commit_and_fetch(
+		const struct ublksrv_ctrl_dev *ctrl_dev,
+		struct io_uring *ring,
+		struct blk_zone *zone_info,
+		uint64_t *cqe_data) {
+
+	struct ublksrv_ctrl_cmd_data data = {
+		.cmd_op = UBLK_CMD_REPORT_ZONES_COMMIT_FETCH,
+		.flags = CTRL_CMD_HAS_BUF,
+		.addr = (__u64)zone_info,
+		.len = sizeof(struct blk_zone),
+	};
+
+	return __ublksrv_ctrl_cmd_with_ring(ctrl_dev, ring, &data, cqe_data);
+}
```

Comment:
All the above report_zones_commit_and_fetch are not good.

Please just define one extra UBLK_IO_OP_ZONE_REPORT
for handling ->report_zone(). And ublksrv_io_desc is 24byte,
which is enough to hold zone_io_report_desc, which
usually just need  start_sector and nr_zones.

Reply:
I think I understand. You would prefer the zone report to go via the ring already in place for the IO requests? I will try to see if I can do this.

The reason I did not do this in the first place is that kernel request to uring request is using the kernel `req->tag`. So if I use the ring to do something that is not IO, I might grab a uring IO that is going to be used by `blk-mq` tag. But I guess I can submit a few extra uring requests for the zone report.

Reply:
> I think I understand. You would prefer the zone report to go via the ring already in place for the IO requests? I will try to see if I can do this.
> 
> The reason I did not do this in the first place is that kernel request to uring request is using the kernel `req->tag`. So if I use the ring to do something that is not IO, I might grab a uring IO that is going to be used by `blk-mq` tag. But I guess I can submit a few extra uring requests for the zone report.

No, when you implement ->report_zones() via blk-mq passthrough
request, it is still blk-mq request, and it is still blk-mq tag.

thanks,

Reply:
> No, when you implement ->report_zones() via blk-mq passthrough
> request, it is still blk-mq request, and it is still blk-mq tag.

I am not a kernel expert, so I might be wrong. This is how I understand `report_zones`: `report_zones()` does not go through the normal block stack. It does not result in a kernel IO request. This is the call stack for `report_zones` ioctl when called on `ublkbX`:

- `blkdev_report_zones_ioctl()` in `blk-zoned.c`
- `blkdev_report_zones()` in `blk-zoned.c`
- `disk->fops->report_zones` which we must implement in `ublk_drv.c`

None of this results in a `blk-mq` request. This is why I state that the `report_zones` ioctl is not implemented through the block layer like open/close/reset/finish zone is. It uses a out-of-band path. Checking `nvme` and `null_blk` I also see no `struct request` being submitted for `report_zones`.

Reply:
->report_zones() is one callback of ublk driver, inside this function,
you call allocate blk-mq requests to forward the request to ublksrv for handling them. Finally ublksrv just fills the report data and 
complete the io command.

That is exactly what nvme ->report_zones() does, please see
nvme_ns_report_zones(). 

That is also why I suggested to define the uapi interface and
post driver code to linux-block for review.